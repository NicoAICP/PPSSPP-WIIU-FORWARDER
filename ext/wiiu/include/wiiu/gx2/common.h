#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
