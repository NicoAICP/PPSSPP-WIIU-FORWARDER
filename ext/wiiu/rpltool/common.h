#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
