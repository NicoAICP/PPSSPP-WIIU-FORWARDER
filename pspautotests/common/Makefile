TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
