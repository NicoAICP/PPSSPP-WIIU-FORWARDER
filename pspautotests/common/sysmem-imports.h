extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();