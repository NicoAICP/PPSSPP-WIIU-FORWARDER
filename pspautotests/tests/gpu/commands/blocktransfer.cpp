#include <common.h>
#include <malloc.h>
#include <pspge.h>
#include <psputils.h>
#include <pspthreadman.h>
#include "commands.h"

unsigned int __attribute__((aligned(16))) dlist1[] = {
	0x00000000, // 0x00 NOP (for GE_CMD_TRANSFERSRC)
	0x00000000, // 0x01 NOP
	0x00000000, // 0x02 NOP (for GE_CMD_TRANSFERSRCPOS)
	0x00000000, // 0x03 NOP (for GE_CMD_TRANSFERDST)
	0x00000000, // 0x04 NOP
	0x00000000, // 0x05 NOP (for GE_CMD_TRANSFERDSTPOS)
	0x00000000, // 0x06 NOP (for GE_CMD_TRANSFERSIZE)
	0xEA000000, // 0x07 TRANSFERSTART
	0x0F000000, // 0x08 FINISH
	0x0C000000, // 0x09 END
};

static const int memsz = 16384 * 32 * sizeof(unsigned int);
unsigned int *mem1;
unsigned int *mem2;

int sameCount(const void *p1, const void *p2) {
	const u8 *b1 = (const u8 *)p1;
	const u8 *b2 = (const u8 *)p2;

	int i = 0;
	while (*b1++ == *b2++)
		++i;

	return i;
}

u32 transferCoords(int cmd, int x, int y) {
	return (cmd << 24) | ((y << 10) & 0x000FFC00) | (x & 0x3FF);
}

u32 transferAddr1(int cmd, void *p) {
	return (cmd << 24) | ((u32)p & 0x00FFFFFF);
}

u32 transferAddr2w(int cmd, void *p, int w) {
	return (cmd << 24) | (((u32)p & 0xFF000000) >> 8) | (w & 0x0000FFFF);
}

void testTransferSize(const char *title, int srcStride, int dstStride, int w, int h, int bpp) {
	memset(mem1, 0xAA, memsz);
	memset(mem2, 0xBB, memsz);

	dlist1[0] = transferAddr1(GE_CMD_TRANSFERSRC, mem1);
	dlist1[1] = transferAddr2w(GE_CMD_TRANSFERSRCW, mem1, srcStride);
	dlist1[2] = transferCoords(GE_CMD_TRANSFERSRCPOS, 0, 0);
	dlist1[3] = transferAddr1(GE_CMD_TRANSFERDST, mem2);
	dlist1[4] = transferAddr2w(GE_CMD_TRANSFERDSTW, mem2, dstStride);
	dlist1[5] = transferCoords(GE_CMD_TRANSFERDSTPOS, 0, 0);
	dlist1[6] = transferCoords(GE_CMD_TRANSFERSIZE, w - 1, h - 1);
	dlist1[7] = (GE_CMD_TRANSFERSTART << 24) | (bpp == 32 ? 1 : 0);

	sceKernelDcacheWritebackAll();
	sceGeListEnQueue(dlist1, dlist1 + 0x20, -1, NULL);
	sceGeDrawSync(0);

	checkpoint("%s: %d", title, sameCount(mem1, mem2));
}

extern "C" int main(int argc, char *argv[]) {
	mem1 = (unsigned int *)memalign(16, memsz);
	mem2 = (unsigned int *)memalign(16, memsz);

	checkpointNext("Block transfers - 16 bit:");
	testTransferSize("0 stride", 0, 0, 1024, 4, 16);
	testTransferSize("0x100 stride", 0x100, 0x100, 1024, 4, 16);
	testTransferSize("0x3F8 stride", 0x3F8, 0x3F8, 1024, 4, 16);
	testTransferSize("0x400 stride", 0x400, 0x400, 1024, 4, 16);
	testTransferSize("0x407 stride", 0x407, 0x407, 1024, 4, 16);
	testTransferSize("0x408 stride", 0x408, 0x408, 1024, 4, 16);
	testTransferSize("0x7FF stride", 0x7FF, 0x7FF, 1024, 4, 16);
	testTransferSize("0xFC00 stride", 0xFC00, 0xFC00, 1024, 4, 16);
	
	checkpointNext("Block transfers - 32 bit:");
	testTransferSize("0 stride", 0, 0, 1024, 4, 32);
	testTransferSize("0x100 stride", 0x100, 0x100, 1024, 4, 32);
	testTransferSize("0x3F8 stride", 0x3F8, 0x3F8, 1024, 4, 32);
	testTransferSize("0x400 stride", 0x400, 0x400, 1024, 4, 32);
	testTransferSize("0x407 stride", 0x407, 0x407, 1024, 4, 32);
	testTransferSize("0x408 stride", 0x408, 0x408, 1024, 4, 32);
	testTransferSize("0x7FF stride", 0x7FF, 0x7FF, 1024, 4, 32);
	testTransferSize("0xFC00 stride", 0xFC00, 0xFC00, 1024, 4, 32);
	
	checkpointNext("Sizes:");
	testTransferSize("1024x512", 0x200, 0x200, 1024, 512, 16);
	testTransferSize("1024x256", 0x200, 0x200, 1024, 256, 16);
	testTransferSize("1024x128", 0x200, 0x200, 1024, 128, 16);
	testTransferSize("1024x64", 0x200, 0x200, 1024, 64, 16);
	testTransferSize("1024x4", 0x200, 0x200, 1024, 4, 16);
	testTransferSize("1024x1", 0x200, 0x200, 1024, 1, 16);
	testTransferSize("1x1", 0x200, 0x200, 1, 1, 16);

	free(mem1);
	free(mem2);

	return 0;
}