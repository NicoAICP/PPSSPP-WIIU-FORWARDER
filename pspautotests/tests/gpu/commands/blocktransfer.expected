[x] Block transfers - 16 bit:
[r] 0 stride: 2048
[r] 0x100 stride: 3584
[r] 0x3F8 stride: 8144
[r] 0x400 stride: 8192
[r] 0x407 stride: 8192
[r] 0x408 stride: 2048
[r] 0x7FF stride: 2048
[r] 0xFC00 stride: 8192

[x] Block transfers - 32 bit:
[r] 0 stride: 4096
[r] 0x100 stride: 7168
[r] 0x3F8 stride: 16288
[r] 0x400 stride: 16384
[r] 0x407 stride: 16384
[r] 0x408 stride: 4096
[r] 0x7FF stride: 4096
[r] 0xFC00 stride: 16384

[x] Sizes:
[r] 1024x512: 525312
[r] 1024x256: 263168
[r] 1024x128: 132096
[r] 1024x64: 66560
[r] 1024x4: 5120
[r] 1024x1: 2048
[x] 1x1: 2
